src/libraries/System.Private.CoreLib/src/System/Collections/Generic/NonRandomizedStringEqualityComparer.cs

```diff
@@ -108,20 +108,23 @@ internal override RandomizedStringEqualityComparer GetRandomizedEqualityComparer
             }
         }
 
-        public static IEqualityComparer<string>? GetStringComparer(object? comparer)
+        public static IEqualityComparer<string>? GetStringComparer(object comparer)
         {
             // Special-case EqualityComparer<string>.Default, StringComparer.Ordinal, and StringComparer.OrdinalIgnoreCase.
             // We use a non-randomized comparer for improved perf, falling back to a randomized comparer if the
             // hash buckets become unbalanced.
-            if (comparer is null)
+
+            if (comparer.Equals(EqualityComparer<string>.Default))
```

Comment:
```suggestion
             if (ReferenceEquals(comparer, EqualityComparer<string>.Default))
```

Reply:
Looking at this again, I do not think my previous feedback was right. This seems to be using `ReferenceEquals` intentionally, to preserve the exact comparer identity as we are swapping between randomized and non-randomized one internally.